src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2533,23 +2533,80 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
         else
             seen_cols[col] = nothing
         end
-        # p might be a column of df so we make sure we unalias
-        if col === p
-            p = copy(p)
-        end
     end
 
-    pp = similar(p)
+    cp = _compile_permutation!(Base.copymutable(p))
+
+    if fun === Base.invpermute!! 
+        pop!(cp)
+        reverse!(cp)
+        push!(cp, 0)
+    end
 
     for (i, col) in enumerate(eachcol(df))
         if !(i in toskip)
-            copyto!(pp, p)
-            fun(col, pp)
+            _cycle_permute!(col, cp)
         end
     end
     return df
 end
 
+# convert a classical permutation to cycle notation, 
+# zeroing the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    firstindex(p) == 1 || 
+        throw(ArgumentError("Permutation vectors must have 1-based indexing"))
+    # this length is sufficient because we do not record 1-cycles,
+    # so the worst case is all 2-cycles.
+    out = similar(p, 3 * length(p) ÷ 2) 
+    out_len = 0
+    start = 0
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start + 1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
+        k = p[start]
+        count -= 1
+        k == start && continue
+        out_len += 1
+        out[out_len] = k
+        p[start] = 0
+        last_k = k
+        checkbounds(Bool, p, k) || throw(ArgumentError("Not a permutation"))
+        out_len += 1
+        k = out[out_len] = p[k]
+        while true
+            count -= 1
+            p[last_k] = 0
+            last_k = k
+            checkbounds(Bool, p, k) || throw(ArgumentError("Not a permutation"))
+            out_len += 1
+            k = out[out_len] = p[k]
+            k == 0 && break
+        end
+    end
+    return resize!(out, out_len)
+end
+
+# permute a vector `v` based on a permutation `p` listed in cycle notation
```

Comment:
Can you please add a comment about the expected format of cycle notation? (i.e. cycles followed by 0 all encoded in a single vector)?